vision/datasets/imageloader.go

```diff
@@ -10,75 +10,104 @@ import (
 	"github.com/wangkuiyi/gotorch/vision/transforms"
 )
 
-// ImageLoader struct
-type ImageLoader struct {
-	r      *tgz.Reader
-	vocab  map[string]int64
-	err    error
-	inputs []torch.Tensor
-	labels []int64
-	trans  *transforms.ComposeTransformer
-	mbSize int
+type sample struct {
+	data  image.Image
+	label int
 }
 
-func (p *ImageLoader) tensorGC() {
-	p.inputs = []torch.Tensor{}
-	p.labels = []int64{}
-	torch.GC()
+// ImageLoader struct
+type ImageLoader struct {
+	r       *tgz.Reader
+	vocab   map[string]int
+	samples chan sample
+	errChan chan error
+	err     error
+	trans1  *transforms.ComposeTransformer // transforms before `ToTensor`
+	trans2  *transforms.ComposeTransformer // transforms after and include `ToTensor`
+	mbSize  int
+	inputs  []torch.Tensor
+	labels  []int64
 }
 
 // NewImageLoader returns an ImageLoader
-func NewImageLoader(fn string, vocab map[string]int64, trans *transforms.ComposeTransformer, mbSize int) (*ImageLoader, error) {
+func NewImageLoader(fn string, vocab map[string]int, trans *transforms.ComposeTransformer, mbSize int) (*ImageLoader, error) {
 	r, e := tgz.OpenFile(fn)
 	if e != nil {
 		return nil, e
 	}
-	return &ImageLoader{
-		r:      r,
-		vocab:  vocab,
-		err:    nil,
-		trans:  trans,
-		mbSize: mbSize,
-	}, nil
+	trans1, trans2 := splitComposeByToTensor(trans)
+	m := &ImageLoader{
+		r:     r,
+		vocab: vocab,
+		// TODO(yancey1989): using a buffered channel to get better performance, but
+		// a bug should be fixed before. c.f. https://github.com/wangkuiyi/gotorch/pull/304#issuecomment-688670165
+		samples: make(chan sample),
```

Comment:
Since FromBlob is using deep copy now, we could have a test on the buffered channel.

Reply:
done